### DISCUSSION TITLE
PerfView Unsure How To Figure Out Performance Issue

### DISCUSSION THREAD

**Discussion by justalexh** [microsoft/perfview#2030] [Q&A]
Hello all,

I'm testing a microservice system on Azure:

- vP2v3 is the service plan
- API testing 

## Issue

- I'm unable to figure out what the actual cause is when it comes to my performance issue, I've tried looking at the threads that are blocked.
- I decided to dig deep into seeing which functions are being called after `SwapContext` happens and when the `BLOCKED_TIME` occurs.
- I have attempted to look between the range for when JIT calls happened as well, but don't have any idea if it really is the cause of performance issues. 
   - Most of the Jit calls are under 0.5ms, but there is a few occasional ones that popup anywhere from 5,000 - 40,000 msec 

## Information: Thread 11296

![Screenshot 2024-05-02 at 5 53 18 PM](https://github.com/microsoft/perfview/assets/130504679/83553738-7588-44d7-9b87-cd41bf4f09b8)

![Screenshot 2024-05-02 at 5 55 45 PM](https://github.com/microsoft/perfview/assets/130504679/e0dbed6e-4c12-48df-9582-513574e0dd45)

- In this case, I often see `ETW` methods being called and I'm not sure if that's due to having multiple monitoring systems logging/profiling the system. 
- I often see`MCGenControlCallBackV2` but can't find much information about it despite it being apparent in the path.

## Information: Thread 8204

- This one has a **LOT** of path's so please bear with me.

![Screenshot 2024-05-02 at 6 22 41 PM](https://github.com/microsoft/perfview/assets/130504679/f05c72ea-ed54-4744-9a95-5fd1280e4968)

### `System.Threading.SpinWait.SpinOnceCore` and `System.Threading.Tasks.Task.SpinWait(int32)`

![Screenshot 2024-05-02 at 6 23 24 PM](https://github.com/microsoft/perfview/assets/130504679/b6890d95-0457-4a08-83ff-f43776cb97a3)

## `CancellationTokenSource.InitializerWithTimer(unsigned int32)`

- We added a cancellation token due to having time outs after 1 minute, but the requests are still going in on the background after it times out.

![Screenshot 2024-05-02 at 6 24 27 PM](https://github.com/microsoft/perfview/assets/130504679/bc0efa77-d459-4409-b4bb-6be49cccca05)
 
![Screenshot 2024-05-02 at 6 23 41 PM](https://github.com/microsoft/perfview/assets/130504679/ab8bb2c6-5c2c-4cd4-827c-7104e955d457)

![image](https://github.com/microsoft/perfview/assets/130504679/597609a8-28a3-4191-9dee-29533ebfd508)

![image](https://github.com/microsoft/perfview/assets/130504679/71afdd4e-c6d8-4cae-8c17-4cce37340e2b)

![image](https://github.com/microsoft/perfview/assets/130504679/8047ee3f-e96b-4913-91e2-c37dc401d286)

## `CoreclrThePreStub` and `System.Net.Http.HttpConnection.Bool`

![Screenshot 2024-05-02 at 6 23 55 PM](https://github.com/microsoft/perfview/assets/130504679/f9d55c88-9cce-41b2-97dc-db2009586eae)

## `HttpConnectionPool;;+<ConnectAsync>.MoveNext()`

![image](https://github.com/microsoft/perfview/assets/130504679/278592b1-0c98-408c-803c-fbd79a66e6db)

## `coreclr!OnCallCountThresholdReachedstub2
![image](https://github.com/microsoft/perfview/assets/130504679/de212b8a-28c9-485d-a6ae-4334853005a2)

## Conclusion

I apologize for all the images, but I'm fairly new to PerfView and Performance so any guidance would help greatly.

**Comment by brianrob** [comment]
@justalexh, can you share a quick description of the setup of your app as well as a description of the performance issue that you're seeing.  Looking at the images is difficult because I can't quite get a full picture - it would be best if you're able to share the trace.  If you would prefer, you can also e-mail it to me at brianrob [at] microsoft.

From the images, the one thing that caught my attention is the time spent in `OnCallCountThresholdReached` - there appears to be a hot lock there, which is unusual.  This would cause delays for methods as they are tiered up (tiered compilation).

**Reply by justalexh** [reply]
@brianrob , thank you for the response. I have emailed you

**Reply by justalexh** [reply]
@brianrob Also, could you explain to me what a hot lock is and why it's unusual and how you noticed it?

What exactly is tiered compilation in this case?

**Reply by brianrob** [reply]
@justalexh, it looks like I did not receive an e-mail from you.  Can you please re-send to brianrob [at] microsoft [dot] com?  Thanks.
